src/routes/coingeckoProxy/index.ts

```diff
@@ -0,0 +1,116 @@
+import fastifyCaching from "@fastify/caching";
+import httpProxy from "@fastify/http-proxy";
+import { FastifyPluginAsync } from "fastify";
+import { ReadableStream } from "stream/web";
+
+const CACHE_TTL = 1 * 60; // 1 min in s
+const SERVER_CACHE_TTL = 1.5 * 60; // 1.5 min in s
+
+type CachedData = {
+  item: {
+    contents: string;
+    headers: Headers;
+  };
+};
+
+function isCachedResponse(cached: any): cached is CachedData {
+  return (
+    cached &&
+    typeof cached === "object" &&
+    "item" in cached &&
+    typeof cached.item === "object" &&
+    "contents" in cached.item
+  );
+}
+
+const coingeckoProxy: FastifyPluginAsync = async (
+  fastify,
+  opts
+): Promise<void> => {
+  // Set up caching
+  fastify.register(fastifyCaching, {
+    expiresIn: CACHE_TTL,
+    serverExpiresIn: SERVER_CACHE_TTL,
+  });
+
+  // Intercepts request and uses cache if found
+  fastify.addHook("onRequest", async function (req, reply) {
+    // First, get rid of the `accept-encoding` header as we cannot deal with it here
+    req.headers["accept-encoding"] = undefined;
+
+    // Then, get the cached request, if any
+    fastify.cache.get(req.url.toLowerCase(), (err, result) => {
+      if (err) {
+        fastify.log.warn(`Error fetching cache for '${req.url}'`, err);
+        return err;
+      }
+      if (isCachedResponse(result)) {
+        fastify.log.info(`onRequest: cache hit for ${req.url}`);
+        reply
+          .headers({ "cache-hit": true, ...result.item.headers })
+          .send(result.item.contents);
+      }
+    });
+  });
+
+  // Intercepts response to store result on cache
+  fastify.addHook("onSend", async function (req, reply, payload) {
```

Comment:
I tried to move the caching phase to after the response is sent to save some time based on this https://github.com/fastify/help/issues/809
But I wasn't able to, couldn't get the payload to be populated for some reason :/